### PR TITLE
Implement handlers for series literal in cudf-polars

### PR DESCRIPTION
**Pull request by wence-**
## Description

A query plan can contain a "literal" polars Series. Often, for example, when calling a contains-like function. To translate these, introduce a new `LiteralColumn` node to capture the concept and add an evaluation rule (converting from arrow).

Since list-dtype Series need the same casting treatment as in dataframe scan case, factor the casting out into a utility, and take the opportunity to handled casting of nested lists correctly.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf_polars/tests/expressions/test_series_literal.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+
+import polars as pl
+
+from cudf_polars.testing.asserts import assert_gpu_result_equal
+
+
+def test_select_series_literal():
```
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf_polars/tests/expressions/test_series_literal.py:10]: Ah, good idea, let me do that.

[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf_polars/tests/expressions/test_series_literal.py]: Done.

File: python/cudf_polars/cudf_polars/utils/dtypes.py [hunk 2]
```diff
@@ -7,13 +7,42 @@
...
+    As well as arrow ``ListType``s, polars can produce
+    ``LargeListType``s and ``FixedSizeListType``s, these are not
+    currently handled by libcudf, so we attempt to cast them all into
+    normal ``ListType``s on the arrow side before consuming the arrow
+    data.
+    """
+    if isinstance(typ, pa.LargeListType):
+        return pa.list_(downcast_arrow_lists(typ.value_type))
+    # We don't have to worry about diving into struct types for now
+    # since those are always NotImplemented before we get here.
```
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf_polars/cudf_polars/utils/dtypes.py:93]: Done.

**Comment by wence-** [comment]
I needed to spike out binop implementations for a bunch of timelike operations because polars has different casting rules to libcudf.

File: python/cudf_polars/tests/expressions/test_literal.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        assert_ir_translation_raises(q, NotImplementedError)
+
+
+def test_select_literal_series():
+    df = pl.LazyFrame({})
+
+    q = df.select(
+        a=pl.Series(["a", "b", "c"], dtype=pl.String()),
+        b=pl.Series([[1, 2], [3], None], dtype=pl.List(pl.UInt16())),
+        c=pl.Series([[[1]], [], [[1, 2, 3, 4]]], dtype=pl.List(pl.List(pl.Float32()))),
```
Comment (vyasr) [inline python/cudf_polars/tests/expressions/test_literal.py:84]: Do you need to wrap these in `pl.lit`, or does that work automatically? Just wondering if that wrapping makes the expression engine treat this any differently.

File: python/cudf_polars/cudf_polars/dsl/expr.py [hunk 4]
```diff
@@ -369,6 +370,29 @@ def do_evaluate(
...
+    def do_evaluate(
+        self,
+        df: DataFrame,
+        *,
+        context: ExecutionContext = ExecutionContext.FRAME,
+        mapping: Mapping[Expr, Column] | None = None,
+    ) -> Column:
+        """Evaluate this expression given a dataframe for context."""
+        # datatype of pyarrow array is correct by construction.
+        return Column(plc.interop.from_arrow(self.value))
```
Comment (vyasr) [inline python/cudf_polars/cudf_polars/dsl/expr.py:393]: Do we have to worry at all about whether the literal is 0, 1, or 2 dimensional, or is that already handled at this point? It seems like it is handled, but I'm not immediately sure how.

**Review by vyasr — APPROVED** [review]
Couple of questions, but I think they're just gaps in my understanding, so approving.

File: python/cudf_polars/tests/expressions/test_literal.py (hunk 3 above)
Comment (wence-) [reply python/cudf_polars/tests/expressions/test_literal.py:84]: No, `select(pl.Series(...))` and `select(lit(pl.Series(...))` desugar to the same intermediate representation.

File: python/cudf_polars/cudf_polars/dsl/expr.py (hunk 4 above)
Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/expr.py:393]: No, we don't have to worry. This is handled in the `translate` phase. If we observe a polars expr containing a `PySeries`, then we produce this node (a `LiteralColumn`). On the other hand if we _don't_ observe a `PySeries` then we produce a `Literal`. The latter corresponds to a scalar whereas this `LiteralColumn` corresponds to a column.

**Pull request merged**